Refine the health check doc. (#4433)
istio/istio.io: content/docs/ops/setup/app-health-check/index.md

keywords: [security,health-check]
---

[Kubernetes liveness and readiness probes](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-probes/)  offers three options for liveness and readiness probes:

1. Command
1. TCP request
1. HTTP request

> This task explains how these approaches work when Istio mutual TLS is enabled.

Reviewer: ```suggestion
This task shows how to use these approaches in Istio with mutual TLS is enabled.
```

> Command and TCP type probes work with Istio regardless whether mutual TLS is enabled or not. HTTP request requires different Istio configuration when

Reviewer: ```suggestion
Command and TCP type probes work with Istio regardless of whether or not mutual TLS is enabled. The HTTP request approach requires different Istio configuration with
```

> Istio mutual TLS is enabled.

Reviewer: ```suggestion
mutual TLS enabled.
```

## Liveness and readiness probes with command option

> In this section, you first configure health checking when mutual TLS is enabled.

Reviewer: I fixed the sentence but the text makes no logical sense to readers. How they can do this is not discussed. Please revise.
```suggestion
First, you need to configure health checking with mutual TLS enabled.
```

    EOF
    {{< /text >}}

> Run this command to deploy the service:

Reviewer: ```suggestion
Run the following command to deploy the service:
```

liveness-6857c8775f-zdv9r        2/2       Running   0           4m
{{< /text >}}

## Liveness and readiness probes with HTTP request option

> This section shows how to configure health checking with the HTTP request option when mutual TLS is enabled.

Reviewer: Why is this here if the readers need to do these steps first? Please ensure the order of the instructions works well after removing the content.

Author: You mean the steps to enable Istio installation with helm flag `sidecarInjectorWebhook.rewriteAppHTTPProbe=true`? Some apporoach like prober rewrite by annotation does not require this helm flag, same for the separate port.

What do you think? Maybe just mention this flag as an option in `Before you Begin`?

liveness-6857c8775f-zdv9r        2/2       Running   0           4m
{{< /text >}}

## Liveness and readiness probes with HTTP request option

This section shows how to configure health checking with the HTTP request option when mutual TLS is enabled.

Kubernetes HTTP health check request is sent from Kubelet, which does not have Istio issued certificate to the `liveness-http` service. So when mutual TLS is enabled, the health check request will fail.

We have two options to solve the problem: probe rewrites and separate ports.

### Probe rewrite

This approach rewrites the application `PodSpec` readiness/liveness probe, such that the probe request will be sent to
[Pilot agent](/docs/reference/commands/pilot-agent/). Pilot agent then redirects the
request to application, and strips the response body only returning the response code.

You have two ways to enable Istio to rewrite the liveness HTTP probes.

> #### Enable via Helm Option

Reviewer 2: add globally ?

Author: Added.

The configuration changes above (by Helm or by the configuration map) effect all Istio app deployments.
{{< /warning >}}

#### Use Annotations on Pod

> <!-- Add samples YAML or kubectl patch? -->

Author: I'm thinking to add sample YAML inlined here, as 

```yaml
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  name: liveness-http
spec:
  template:
    metadata:
      labels:
        app: liveness-http
        version: v1
      annotations:
        sidecar.istio.io/rewriteAppHTTPProbers: "true"
    spec:
      containers:
      - name: liveness-http
        image: docker.io/istio/health:example
        ports:
        - containerPort: 8001
        livenessProbe:
          httpGet:
            path: /foo
            port: 8001
          initialDelaySeconds: 5
          periodSeconds: 5
```

And let people to do kubectl apply, as opposed to just mention the annotation. I think that makes things easier to follow? wtdy? @rcaballeromx

Reviewer 2: code snippet to copy is best imho

Author: Added.

The configuration changes above (by Helm or by the configuration map) effect all Istio app deployments.
{{< /warning >}}

#### Use Annotations on Pod

> 

Reviewer 2: I think this option should go first.  It is most likely users will use annotation before enable it globally.

Author: Maybe we should even focus on document this approach and then in the end mention the Helm flag can turn it on globally as a side note? That makes the doc even simpler.

Author: This might suits well in the follow-up improvements.
